File: fb/235.lowest-common-ancestor-of-a-binary-search-tree.cpp
// Time O(n)
// Space O(1)
class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        int pVal = p->val;
        int qVal = q->val;
        TreeNode* node = root;
        while(node) {
            int parentVal = node->val;
            if(pVal > parentVal && qVal > parentVal) {
                node = node->right;
            } else if (pVal < parentVal && qVal < parentVal) {
                node = node->left;
            } else {
                return node;
            }
        }
        return nullptr;
    }
};

// Time O(n)
// Space O(n)
class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        if(!root || root == p || root == q) return root;
        if(root->val < p->val && root->val < q->val)
            return lowestCommonAncestor(root->right, p, q);
        else if(root->val > p->val && root->val > q->val)
            return lowestCommonAncestor(root->left, p, q);
        return root;
    }
};

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

File: fb/235.lowest-common-ancestor-of-a-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        # parentVal = root.val
        # pVal = p.val
        # qVal = q.val
        # if pVal < parentVal and qVal < parentVal:
        #     return self.lowestCommonAncestor(root.left, p, q)
        # elif pVal > parentVal and qVal > parentVal:
        #     return self.lowestCommonAncestor(root.right, p, q)
        # else:
        #     return root
        
        pVal = p.val
        qVal = q.val
        node = root
        while node:
            parentVal = node.val
            if pVal < parentVal and qVal < parentVal:
                node = node.left
            elif pVal > parentVal and qVal > parentVal:
                node = node.right
            else:
                return node
